### PR TITLE
add option to remember selected layout per window

### PR DESCRIPTION
**Pull request by krydos**
The PR for issue #2

### CODE REVIEW

File: kbdcfg.lua [hunk 1]
```diff
@@ -105,6 +107,18 @@ local function factory(args)
     return kbdcfg
 end
 
+function kbd_client_update(c)
```
Comment (echuraev) [inline kbdcfg.lua]: I think it would be better to move this function before factory function.

File: README.md [hunk 2]
```diff
@@ -102,6 +102,15 @@ Note, that you should pass a valid input method name to `fcitx-remote`
 command. The last layout in the example uses the
 [Mozc](https://github.com/google/mozc) as an input method for Japanese input.
 
+## How to keep chosen layout per window
+
+If you wish to remember selected layout per window set `remember_layout` variable of your `kbdcfg` like so:
+
+```
+kbdcfg.remember_layout = true
```
Comment (echuraev) [inline README.md]: I think it would be better to add this variable to factory function.
And then we will be able to set this variable during creation kbdcfg object, e.g.:
`local kbdcfg = keyboard_layout.kbdcfg({type="gui", remember_layout=true})`

**Review by echuraev — CHANGES_REQUESTED** [review]

File: kbdcfg.lua [hunk 3]
```diff
@@ -26,11 +26,9 @@ function kbdcfg.switch_by_name(name)
 end
 
 function kbdcfg.switch(layout)
-    for i, current in ipairs(kbdcfg.layouts) do
```
Comment (echuraev) [inline kbdcfg.lua]: Why did you remove this code? Are you sure that your code will work well in any cases?

Comment (krydos) [reply kbdcfg.lua]: @echuraev no no, this is exactly what I was think about at today's morning :D I've just blindly replaced this part with the new one and it was working in my test case but definitely should break something. Give me some time I'll update the PR

**Comment by krydos** [comment]
@echuraev I've changed `remember_layout` variable to be set via factory function. Looks much cleaner, thanks for pointing to it. 

`kbd_client_update` function moved above of factory function.

I didn't really get the code I've removed from `switch` method, and also didn't get why it works without it... 
Anyway I've returned it back and since I didn't get it I've wrapped it with `if/else` block so if `remember_layout` is `false` then the plugin should work as before.

File: kbdcfg.lua [hunk 4]
```diff
@@ -26,11 +26,19 @@ function kbdcfg.switch_by_name(name)
...
+
+    if kbdcfg.remember_layout then
+      if client.focus then
+        local layout = kbdcfg.layouts[kbdcfg.current]
+        client.focus.kbd_layout = layout.name
+      end
+    else
+      for i, current in ipairs(kbdcfg.layouts) do
```
Comment (echuraev) [inline kbdcfg.lua]: In this code you just we just looking for index of new layout and set the value to `kbdcfg.current` variable. I think that it's better to run this loop before check of client in focus.
You are right this code is not clear enough. I wrote this code about 8 months ago and may be it is a time to make small refactoring for the code after your commit.
Tomorrow, I'll check if it will work correctly. Could you please remove the if condition on line 30 and move this loop before checking client.focus? It would be nice if you could test it on your configuration. Please, let me know if it won't work.

File: kbdcfg.lua [hunk 5]
```diff
@@ -95,6 +119,7 @@ local function factory(args)
     kbdcfg.current               = args.current or 1
     kbdcfg.menu                  = nil
     kbdcfg.type                  = args.type or "tui"
+    kbdcfg.remember_layout          = args.remember_layout or false
```
Comment (echuraev) [inline kbdcfg.lua]: Could you please align this initialisation with others (just remove 3 spaces before '=')?

File: kbdcfg.lua [hunk 6]
```diff
@@ -26,11 +26,19 @@ function kbdcfg.switch_by_name(name)
 end
 
 function kbdcfg.switch(layout)
-    for i, current in ipairs(kbdcfg.layouts) do
+
```
Comment (echuraev) [inline kbdcfg.lua]: Please, remove empty line

File: kbdcfg.lua [hunk 7]
```diff
@@ -26,11 +26,19 @@ function kbdcfg.switch_by_name(name)
...
+    if kbdcfg.remember_layout then
+      if client.focus then
```
Comment (echuraev) [inline kbdcfg.lua]: Indents should be equal 4 to spaces.

File: kbdcfg.lua [hunk 8]
```diff
@@ -59,6 +67,18 @@ function kbdcfg.add_additional_layout(name, label, subcmd)
     table.insert(kbdcfg.additional_layouts, layout)
 end
 
+function kbd_client_update(c)
```
Comment (echuraev) [inline kbdcfg.lua]: Indents should be equal 4 to spaces.

File: kbdcfg.lua [hunk 9]
```diff
@@ -81,6 +101,10 @@ function kbdcfg.bind()
         kbdcfg.widget = wibox.widget.imagebox()
     end
 
+    if kbdcfg.remember_layout then
+      client.connect_signal("focus", kbd_client_update)
```
Comment (echuraev) [inline kbdcfg.lua]: Indent

**Review by echuraev — CHANGES_REQUESTED** [review]

**Comment by krydos** [comment]
Oh, crap. I've really messed up with indentation, sorry about that. 
I'll try to move for loop for kbdcfg.current as you have suggested and let you know about the result

**Comment by krydos** [comment]
hi @echuraev 
All your comments are addressed now. The plugin seems working on my configuration with `remember_layout` option set and without it. 

I hope it will work for you as well

**Comment by krydos** [comment]
Let me know if you think I can make something better here

File: kbdcfg.lua [hunk 10]
```diff
@@ -59,6 +64,18 @@ function kbdcfg.add_additional_layout(name, label, subcmd)
     table.insert(kbdcfg.additional_layouts, layout)
 end
 
+function kbd_client_update(c)
+    if not c then
+        return
+    end
+
+    if not c.kbd_layout then
+        c.kbd_layout = kbdcfg.layouts[1].name
```
Comment (echuraev) [inline kbdcfg.lua]: May be it is better to use here `kbdcfg.current` instead of constant 1. In this case we will set current layout for every new windows.

**Review by echuraev — APPROVED** [review]

**Comment by echuraev** [comment]
@KryDos Thank you very much for your PR. It looks great.
I'll test it on my configuration and next I'll merge this PR to master. If you are able to apply my last comment it would be nice. But I can do after merging the PR during refactoring.

**Comment by krydos** [comment]
@echuraev yeah, it's definitely  make sense. I've changed magic constant to `kbdcfg.current`.

**Comment by krydos** [comment]
Sorry, I've blindly copy pasted everything and didn't pay much attention on how it actually works.

**Comment by krydos** [comment]
Hm... just reconsider it... I don't even know what is better. Maybe magic constant is good (we can call it like `default_layout`). Since in case of using `kbdcfg.current` every new window will use previously selected kbd layout. 

Not sure if it's good idea. In my case it's better to use default layout for new windows (which is EN in my case). 

@echuraev what do you think?

**Comment by echuraev** [comment]
@KryDos Yes, may be you are right. Usually I'm not using this feature with remembering layouts, so, for me it is more intuitive that we will set current layout for a new window. But if you think that it could be better to set default layout for new windows, so we can try to do it. Anyway in the future it is possible to add different variants which layout should be set. Now I don't see any reasons to keep my approach. 
I'lll see on the requests from users of the widget.

Could you please add variable with value of default layout index to factory function, update documentation that user can set the index of his default layout and replace all uses of index 1 to the new variable? I think that it should be enough.

**Comment by krydos** [comment]
@echuraev just made a change. Added `default_layout_index` variable that can be set in factory. Sure there are some places to improve it but it looks pretty usable for now. 

Thanks a lot for the help here.

File: kbdcfg.lua [hunk 11]
```diff
@@ -95,6 +116,8 @@ local function factory(args)
     kbdcfg.current               = args.current or 1
     kbdcfg.menu                  = nil
     kbdcfg.type                  = args.type or "tui"
+    kbdcfg.remember_layout       = args.remember_layout or false
+    kbdcfg.default_layout_index  = args.default_layout_index or 1
```
Comment (echuraev) [inline kbdcfg.lua]: Could you please move it before `kbdcfg.current` and use this value as a default value for `kbdcfg.current`?

File: README.md [hunk 12]
```diff
@@ -102,6 +102,22 @@ Note, that you should pass a valid input method name to `fcitx-remote`
...
+local kbdcfg = keyboard_layout.kbdcfg({remember_layout=true})
+```
+
+You can also set default layout for new windows with `default_layout_index` variable like that:
+
+```
+local kbdcfg = keyboard_layout.kbdcfg({remember_layout=true, default_layout_index=1})
+```
+
+Index depends on your `kbdcfg.add_primary_layout` calls.
```
Comment (echuraev) [inline README.md:119]: I suggest to create separate section for `default_layout_index` because we can use it to set `kbdcfg.current` and in this case it is just the index of default layout and it doesn't depends on `kbdcfg.add_primary_layout`.

And also, it is necessary to add something like that: "Index of layout is set in the order of adding layouts".

I'm not sure that it is a good idea to create a lot of sections for options, functions, usage and etc. I'll think how organize documentation in better way. @KryDos May be do you have any inputs?

**Comment by krydos** [comment]
I've changed `kbdcfg.current` to use value of `default_layout_index`. I'm not really sure at this point how to check if value assigned to `default_layout_index` is correct because assigning is happening after initialization. 

As for doc organization. I don't have much thoughts about it but maybe instead of `How to keep chosen layout per window` let's have
something like `Available options` header with list of all options and brief description, like:

```
Available options:

* cmd  - command to change layout
* type - set type of widget
* etc...
```

File: kbdcfg.lua [hunk 13]
```diff
@@ -81,6 +98,10 @@ function kbdcfg.bind()
         kbdcfg.widget = wibox.widget.imagebox()
     end
 
+    if kbdcfg.remember_layout then
```
Comment (echuraev) [inline kbdcfg.lua]: In this function you already know the number of layouts and you can add a check which value user have set for `kbdcfg.default_layout_index`. I suggest if index is incorrect we should set it equal to 1. Also it is necessary to update kbdcfg.current in case of wrong value.
You can add check before this if condition.

**Comment by echuraev** [comment]
Hi @KryDos 

Sorry for delay in response. I added comment about how we can check the value of `kbdcfg.default_layout_index`. 

Thank you for your point about documentation. I think that now we can create a separate section for the new option. And after your commit I'll reorgonize the whole documentation. I'd like to add TOC and also create spetial section with widget options and with features of this widget.

**Comment by krydos** [comment]
@echuraev sorry for my a bit slow reaction. Women's day takes more time than learning Lua :rofl: 

I've added the check BUT I had to remove `kbdcfg.bind()` call from the factory since when it called from factory the `kbdcfg.layouts` has nothing in it and `default_layout_index` is always out of range and `default_layout_index` is set to 1 (default value). So on the next call of bind we loose original value of `default_layout_index`.

Based on what I understand the `kbdcfg.bind` call from factory method is not required since documentation requires to call it in rc.lua.

**Comment by echuraev** [comment]
@KryDos Thank you very much for your contribution in the widget. It looks good. 

You are right, I don't remember why I added `kbdcfg.bind` function to the factory. It was necessary to remove this call. Thank you!

I'll update the docs.
